### ISSUE TITLE
Why I don't do Domain Driven Design

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#7379]
Why I don't do Domain Driven Design<br>
<br>
https://ift.tt/2ZIavio<br>
<br>
Tony Marston<br>
<br>
<div>
<div>
<div>
<h1>Why I don't do Domain Driven Design</h1>
<div>
<span>Posted on 1st March 2018</span> <span>by Tony Marston</span>
<p>Amended on 2nd March 2019</p>
</div>
<dl>
<dt><a href="https://www.tonymarston.net/php-mysql/dont-do-domain-driven-design.html#introduction">Introduction</a></dt>
<dt><a href="https://www.tonymarston.net/php-mysql/dont-do-domain-driven-design.html#what.is.a.domain">What is a "domain"?</a></dt>
<dt><a href="https://www.tonymarston.net/php-mysql/dont-do-domain-driven-design.html#differences.and.similarities">The biggest differences with the most similarities</a></dt>
<dt><a href="https://www.tonymarston.net/php-mysql/dont-do-domain-driven-design.html#software.or.database">Which has priority, the software or the database?</a></dt>
<dt><a href="https://www.tonymarston.net/php-mysql/dont-do-domain-driven-design.html#build.a.library">Build a library of reusable components</a></dt>
<dt><a href="https://www.tonymarston.net/php-mysql/dont-do-domain-driven-design.html#library.then.framework">Turning a library into a framework</a></dt>
<dt><a href="https://www.tonymarston.net/php-mysql/dont-do-domain-driven-design.html#use.the.framework">Use the framework to build components</a></dt>
<dt><a href="https://www.tonymarston.net/php-mysql/dont-do-domain-driven-design.html#tasks.not.methods">Use Cases require tasks, not methods</a></dt>
<dt><a href="https://www.tonymarston.net/php-mysql/dont-do-domain-driven-design.html#optimise.the.ui">Optimise the User Interface</a></dt>
<dt><a href="https://www.tonymarston.net/php-mysql/dont-do-domain-driven-design.html#business.rules">How can specific business rules be inserted into the application?</a></dt>
<dt><a href="https://www.tonymarston.net/php-mysql/dont-do-domain-driven-design.html#subclassing">How can the same method execute different code for different tasks?</a></dt>
<dt><a href="https://www.tonymarston.net/php-mysql/dont-do-domain-driven-design.html#conclusion">Conclusion</a></dt>
<dt><a href="https://www.tonymarston.net/php-mysql/dont-do-domain-driven-design.html#references">References</a></dt>
<dt><a href="https://www.tonymarston.net/php-mysql/dont-do-domain-driven-design.html#history">Amendment History</a></dt>
<dt><a href="https://www.tonymarston.net/php-mysql/dont-do-domain-driven-design.html#comments">Comments</a></dt>
</dl>
<p>What is Domain Driven Design (DDD)? If you read <a href="https://en.wikipedia.org/wiki/Domain-driven_design">this Wikipedia article</a> you will see that it talks about DDD consisting of a number of high-level concepts and practices. It refers to a ubiquitous language meaning that the domain model should form a common language given by domain experts for describing system requirements. It talks about breaking the domain into its entities, value objects, aggregates, events, services, repositories and factories. The primary focus of this process is the domain logic, the business rules for that domain. Any logic which is not specific to a particular domain, such as communicating with the database or the user interface, is left until last as it is considered to be nothing more than an implementation detail.</p>
<p>I have been designing and building enterprise applications for nearly 4 decades, and this is NOT the way that I do it. The design and implementation of any software application has the following phases:</p>
<ul>
<li>
<a name="logical.design" id="logical.design"></a><b>Logical Design</b> - This is produced by analysts as the first response to the system requirements. It requires the identification of each Event (or <b>Use Case</b>, <b>User Transaction</b> or <b>Task</b>) so that the inputs and outputs can be designed, either in the form of HTML pages, PDF documents or CSV files, as well as the business rules which need to be followed. It also begins to identify the logical database design. The client is given a walk-through of this Logical Design in order to verify that it meets all his requirements, to clear up any misunderstandings and ensure that nothing is missing.</li>
<li>
<a name="physical.design" id="physical.design"></a><b>Physical Design</b> - This can only be produced after the logical design has been signed off by the client. It identifies what hardware will be needed, what software tools will be needed (such as the DBMS, programming language, framework and libraries), what programming resources will be needed, in what order the software components will be built, and the Test Plan. This is also the stage at which detailed program specifications are written so the the programmers know precisely what they are supposed to build. The logical database design is then turned into a physical database using the chosen DBMS.</li>
<li>
<a name="implementation" id="implementation"></a><b>Implementation</b> - This is where the individual components are coded and then assembled into the final product. This is where the classes and their methods which will exists in the Business layer are built, the business rules are coded, and the individual events/tasks are built and tested. The chosen framework may provide a certain number of reusable components, so the more it provides the less will have to be developed separately.</li>
</ul>
<p>I analyse the business requirements in order to produce a logical design which identifies <i>what</i> needs to be done, and after this has been verified by the client I produce a physical design which identifies <i>how</i> it will be done. Then and only then can I begin the implementation. Some developers like to design the software components first and leave the database design till last as they consider it to be nothing more than an "implementation detail", but this usually produces what is known as an <a href="https://en.wikipedia.org/wiki/Object-relational_impedance_mismatch">Object-Relational Impedence Mismatch</a> which causes its own set of problems. To me it is the business logic which is the implementation detail and which can be left until last, so I prefer to design and build my database first, then create my software components to match the database structure, thus avoiding the possibility of any mismatch. This produces a separate class for each database table, but I use a framework which generates these classes for me. I recognised decades ago that in a database application each event/task will interact with one or more database tables and that the only operations which can be performed on a database table are Create, Read, Update and Delete. I also recognised that different events/tasks can sometimes have identical behaviour with the only difference being the database table(s) on which they operate, so I encapsulated this behaviour into a series of reusable standard templates known as <a href="https://www.tonymarston.net/php-mysql/dialog-types.html">Transaction Patterns</a>. To create a new event/task I simply join a <a href="https://www.tonymarston.net/php-mysql/dialog-types.html">Transaction Pattern</a> with one or more table classes, press a button, and the framework automatically creates a working task for me. That standard template will provide the basic functionality so that I can run the transaction and communicate with the database, and I add in the specific business rules afterwards by filling the relevant <a href="https://www.tonymarston.net/php-mysql/radicore-hooks.html">hook methods</a> with the necessary code.</p>
<p>It was not until 2002 that I started using a language that had OO capabilities, and at that time the notions of <a href="https://en.wikipedia.org/wiki/Object-oriented_design">Object Oriented Design (OOD)</a> and <a href="https://en.wikipedia.org/wiki/Domain-driven_design">Domain Driven Design (DDD)</a> had either not yet been formulated, or were in their infancy and were not widely publicised nor generally accepted in the programming community. All that I knew was that <a href="https://www.tonymarston.net/php-mysql/what-is-oop.html#what.oop.is">Object Oriented Programming (OOP)</a> was a programming technique that was similar to procedural programming but with the addition of <a href="https://www.tonymarston.net/php-mysql/oop-for-heretics.html#encapsulation">encapsulation</a>, <a href="https://www.tonymarston.net/php-mysql/oop-for-heretics.html#inheritance">inheritance</a> and <a href="https://www.tonymarston.net/php-mysql/oop-for-heretics.html#polymorphism">polymorphism</a> which could be used to increase code reuse and decrease code maintenance. I say <i>could</i> because the use of OO techniques does not automatically guarantee that the code will be <i>better</i>, the only guarantee is that it will be <i>different</i>. The result of using OOP is still down to the skill of the individual programmer. If you read any articles on how to make effective use of these three fundamental concepts you will see such ideas as <a href="https://www.tonymarston.net/php-mysql/oop-for-heretics.html#coupling">coupling</a>, <a href="https://www.tonymarston.net/php-mysql/oop-for-heretics.html#cohesion">cohesion</a> and <a href="https://www.tonymarston.net/php-mysql/oop-for-heretics.html#dependency">dependency</a> being mentioned. These ideas existed in other languages before OOP was invented, so they are not unique to OOP. OOP itself is not a programming technique which is totally different from what came before, it uses the same basic principles, but adds encapsulation, inheritance and polymorphism into the mix. This is discussed in more detail in <a href="https://www.tonymarston.net/php-mysql/difference-between-procedural-and-oo-programming.html">What is the difference between Procedural and OO programming?</a></p>
<p>While Domain Driven Design concentrates on the unique aspects of each particular domain, which is fine for the <a href="https://www.tonymarston.net/php-mysql/dont-do-domain-driven-design.html#logical.design">Logical Design</a>, the novice programmer may also assume that the <a href="https://www.tonymarston.net/php-mysql/dont-do-domain-driven-design.html#physical.design">Physical Design</a> and the eventual <a href="https://www.tonymarston.net/php-mysql/dont-do-domain-driven-design.html#implementation">Implementation</a> should also be based around these unique aspects, but this again is an area where I choose to be different. Anyone who has ever built a number of user transactions will be able to tell you that writing the code to implement the business rules is only a small part of the total code that will actually be required to execute those business rules. There is a lot of code which handles the communication with the client, the construction of HTML, CSV and PDF documents, the validation of user input, and the communication with the database. This is code which is not part of the domain, but which has to be written in order to get the application to work. Experienced programmers should be able to look at a number of different user transactions and see a number of similarities with this non-domain code. Not only will there be similarities in this non-domain code within the user transactions for an individual domain but these similarities will also appear in the transactions for other domains. For example, if your application has several domains which each communicates with a relation database, then the structure of the code which performs this communication will be identical regardless of the particular domain. While other programmers are being taught to start with the domain logic and add in the non-domain logic afterwards I have found great benefit in doing the exact opposite - I start with the non-domain logic and add in the domain logic afterwards.</p>
<p>According to <a href="https://airbrake.io/blog/software-design/domain-driven-design">Domain-Driven Design - What is it and how do you use it?</a> the term "domain" can be defined as:</p>
<blockquote>A sphere of knowledge and activity around which the application logic revolves.</blockquote>
<p>That's all well and good, but how do you recognise when you are developing software which deals with more than one domain? As an example I have developed a large ERP application which currently contains the following subsystems:</p>
<a name="multiple.domains" id="multiple.domains"></a>
<ol>
<li>Role Based Access Control - users, events/tasks, permissions, menus.</li>
<li>Audit - logs all database changes.</li>
<li>Workflow - activity based workflows.</li>
<li>Data Dictionary - class generation and transaction generation.</li>
<li>Party - people and organisations, customers and suppliers.</li>
<li>Product - products, components, features, prices.</li>
<li>Order - sales orders, purchase orders, transfer orders.</li>
<li>Invoice - sales invoices, purchase invoices, debit notes and credit notes.</li>
<li>Inventory - products in inventory, issuances, receipts, facilities, containers, lots.</li>
<li>Shipment - outbound shipments, inbound shipments, customer shipments, supplier shipments.</li>
<li>Request - requirements, requests and quotations.</li>
<li>Survey - surveys and questionnaires.</li>
<li>Content - miscellaneous content.</li>
<li>Work Effort - timesheets and expenses.</li>
<li>Blockchain - configuration and administration.</li>
<li>Personnel - strategic HR, performance reviews.</li>
</ol>
<p>This is not a complete list as work on several new subsystems is ongoing.</p>
<p>Each of these subsystems has its own database and its own directory structure for all its scripts. They are integrated by virtue of the fact that data can be shared. For example, the ORDER subsystem references the PARTY subsystem for details of customers or suppliers, and references the PRODUCT subsystem for details of products and services which can be bought or sold. It would be very inefficient if the ORDER subsystem had its own tables for customers, suppliers and products.</p>
<p>The <a name="domain.differences" id="domain.differences">differences</a> between each of these subsystems/domains can be categorised as follows:</p>
<ul>
<li>Each has its own database with its own collection of tables, and each table has a unique arrangement of column names.</li>
<li>Each has its own set of events/tasks which perform different operations on different tables and create different outputs for the user.</li>
<li>Each has its own set of methods containing unique pieces of code to execute business rules.</li>
</ul>
<p>When a novice programmer first examines different domains within an enterprise application a casual glance will immediately highlight the differences - the inputs are different, the outputs are different, the database is different, the user transactions (use cases) are different, and the business rules are different. That novice will then assume that because each domain is so different from the others that it will require a totally different design and therefore totally different code to implement that design. However, when you have built as many database applications as I have you will quickly learn that the differences actually hide a number of similarities which provide opportunities for significant amounts of reusable and sharable code. An experienced programmer will tell you that the more reusable code you have then the less code you have to write. This leads to shorter development timescales, lower costs, quicker <a href="https://en.wikipedia.org/wiki/Time_to_market">time-to-market</a> (TTM) and easier maintenance.</p>
<p>But where are these similarities I hear you ask? Where are the opportunities for reusable code? As noted previously I have developed a large ERP application which contains <a href="https://www.tonymarston.net/php-mysql/dont-do-domain-driven-design.html#multiple.domains">a number of different domains</a> which have their <a href="https://www.tonymarston.net/php-mysql/dont-do-domain-driven-design.html#domain.differences">differences</a>. However, an experienced developer can look at this list of differences and see a number of similarities. If you read <a href="https://www.tonymarston.net/php-mysql/anatomy-of-an-enterprise-application.html">Anatomy of an Enterprise Application</a> you will see that, regardless of the domain, each user transaction follows the same pattern - there is a User Interface (UI) at the front end, a database at the back end, and software in the middle which transfers data between the two, and which applies the business rules. There are basically two ways in which you can view this problem, and your design of the solution will be heavily influences by which view you take:</p>
<ul>
<li>You have a series of unique business rules which require program code, and which incidentally require the use of HTML forms at the front end and SQL queries at the back end. You start by writing the code which executes these business rules, then you add in the code which handles the HTML and SQL interaction using what is known as <a href="https://en.wikipedia.org/wiki/Boilerplate_code">boilerplate code</a> as it is usually quite repetitious and boring.</li>
<li>You have code which deals with HTML at the front end, SQL queries at the back end, and which incidentally executes some business rules in the middle. You start by building a working skeleton of a user transaction from a template or pattern, then you flesh it out by adding in the relevant business rules later. This method is much faster and therefore more cost-effective, but it does require a significant amount of up-front investment to create the necessary templates or patterns. When you do what I do and build dozens of domains with thousands of user transactions this investment is definitely worth it.</li>
</ul>
<p>A common way to deal with these three areas is to use an architectural pattern known as the <a href="https://www.tonymarston.net/php-mysql/3-tier-architecture.html">3-Tier Architecture (3TA)</a> which has a Presentation layer for the User Interface (UI), a Business layer (also known as the Domain layer) for all the business/domain logic, and a Data Access layer for communicating with the physical database. Another common pattern is the <a href="https://www.tonymarston.net/php-mysql/model-view-controller.html">Model-View-Controller (MVC)</a> which I have merged with 3TA by splitting the Presentation layer into a <a href="https://www.tonymarston.net/php-mysql/model-view-controller.html#controller">Controller</a> and a <a href="https://www.tonymarston.net/php-mysql/model-view-controller.html#view">View</a>, with the <a href="https://www.tonymarston.net/php-mysql/model-view-controller.html#model">Model</a> being in the Business layer. This produces an application architecture which is shown in <a href="https://www.tonymarston.net/php-mysql/dont-do-domain-driven-design.html#figure1">Figure 1</a>:</p>
<p>The following areas have similar characteristics which provide opportunities for reusable code:</p>
<ul>
<li>
<b>Inputs</b>. In a web application all inputs come from an HTML form as an HTTP request which uses either the GET or POST method. In the PHP language the data for each of these methods is presented to the code as an associative array. It does not matter what data is contained in this array, it is a standard structure which can be accessed and manipulated using standard code. The HTTP request is passed to an initial <a href="https://www.tonymarston.net/php-mysql/infrastructure.html#component.script">component script</a> which sends the request to a <a href="https://www.tonymarston.net/php-mysql/infrastructure.html#controller.script">Controller</a> which then translates the request into specific method calls on a nominated <a href="https://www.tonymarston.net/php-mysql/infrastructure.html#table.class">Model</a> (or Models). When the Model has finished processing the request, which may or may not involve communicating the <a href="https://www.tonymarston.net/php-mysql/infrastructure.html#dml.class">Data Access Object</a>, it gives its raw data to the <a href="https://www.tonymarston.net/php-mysql/infrastructure.html#view.object">View</a> which then produces the output which is returned to the user.</li>
<li>
<b>Outputs</b>. The most common form of output in a web application is an HTML form, and the only difference between one HTML form and another is its structure and/or its content. In a large application with thousands of forms you may be surprised to know that the number of different structures is actually quite small. I have produced a library of just over a dozen <a href="https://www.tonymarston.net/php-mysql/infrastructure.html#xsl.stylesheet">XSL stylesheets</a> which identify the possible structures, and by using common code to put the application data into an <a href="https://www.tonymarston.net/php-mysql/infrastructure.html#xml.file">XML document</a> I can then perform an <a href="https://www.tonymarston.net/php-mysql/infrastructure.html#xsl.transformation">XSL Transformation</a> which produces the <a href="https://www.tonymarston.net/php-mysql/infrastructure.html#html.output">HTML</a> output. This is all handled by a single <a href="https://www.tonymarston.net/php-mysql/infrastructure.html#view.object">View</a> component which is aided by a small <a href="https://www.tonymarston.net/php-mysql/infrastructure.html#screen.structure">screen structure script</a> which does nothing but identify which piece of application data goes where.
<p>I have a similar process to produce PDF output. I have a single <a href="https://www.tonymarston.net/php-mysql/output-to-pdf.html">PDF component</a> which is aided by a small <a href="https://www.tonymarston.net/php-mysql/output-to-pdf.html#report.structure.file">report structure script</a> which does nothing but identify which piece of data goes where.</p>
<p>Creating CSV output is the easiest - simply extract the array of data from the <a href="https://www.tonymarston.net/php-mysql/infrastructure.html#table.class">Model</a> component and write it to a disk file. I do not have to know what data this array contains, so I can extract different arrays of data from any number of different objects and the process of creating a CSV file from this data is exactly the same.</p>
</li>
<li>
<b>Databases</b>. While it is true that every different domain uses totally different sets of data, when it is stored in a relational database this data is contained in objects called tables. While each table has a unique name and a unique structure in the form of a particular collection of columns, this is where the differences end. Each column has a data type which comes from a restricted list, a size which can be varied, and a NULL or NOT NULL attribute which indicates whether it is optional or required. Regardless of its structure, every table in the database is accessed by the same set of SQL operations which are <a href="https://en.wikipedia.org/wiki/Create,_read,_update_and_delete">Create, Read, Update and Delete</a>, or CRUD for short. Each of these operations is constructed as a query string which has a standard format, therefore can quite easily be handled in a sharable <a href="https://www.tonymarston.net/php-mysql/infrastructure.html#dml.class">Data Access Object</a>. I have a separate version of this object for each DBMS that I support.</li>
</ul>
<p>Another common mistake that a novice programmer often makes is to consider each event/task in a domain as a series of business rules which incidentally touch the database. I learned a long time ago that the best approach is to take the opposite view - each event/task is mainly concerned with accessing one or more database tables using one or more of the standard <a href="https://en.wikipedia.org/wiki/Create,_read,_update_and_delete">CRUD</a> operations, and only incidentally executing the business rules. This is why I always start with the standard code and add in the business rules later. In this way I have been able to incorporate a large number of instances of the <a href="https://www.tonymarston.net/php-mysql/template-method-pattern-as-a-framework.html">Template Method Pattern</a> into my framework. This utilises an <a href="https://www.tonymarston.net/php-mysql/infrastructure.html#generic.table.class">abstract class</a> which contains the numerous template methods and invariant methods which contain the standard code, and a collection of <a href="https://www.tonymarston.net/php-mysql/radicore-hooks.html">"hook" methods</a> which can have different implementations in each <a href="https://www.tonymarston.net/php-mysql/infrastructure.html#table.class">subclass</a> in order to handle the business rules which are specific to that subclass.</p>
<p>When you look at the four types of component in <a href="https://www.tonymarston.net/php-mysql/dont-do-domain-driven-design.html#figure1">Figure 1</a> you should now be able to see how I provide them in my code:</p>
<ul>
<li>
<b>Controllers</b> - I have one pre-written and sharable controller for each of the <a href="https://www.tonymarston.net/php-mysql/dialog-types.html">Transaction Patterns</a> which I have recognised.</li>
<li>
<b>Views</b> - I have one pre-written and sharable object for each of the three types of output - HTML, PDF and CSV.</li>
<li>
<b>Models</b> - These are constructed separately once the domain database has been built, one class per table. This process has been automated in that the <a href="https://www.tonymarston.net/php-mysql/data-dictionary.html#table.class.inc">class file</a>, which inherits large volumes of standard code from a single <a href="https://www.tonymarston.net/php-mysql/infrastructure.html#generic.table.class">abstract table class</a>, and <a href="https://www.tonymarston.net/php-mysql/data-dictionary.html#table.dict.inc">structure file</a> are produced by operations within my <a href="https://www.tonymarston.net/php-mysql/data-dictionary.html">Data Dictionary</a>.</li>
<li>
<b>Data Access Object</b> - I have one pre-written and sharable object for each DBMS that I support.</li>
</ul>
<p>Each user transaction - and there may be hundreds, or even thousands - will therefore contain a mixture of code which falls into one of the following categories:</p>
<ul>
<li>
<b>Domain-agnostic</b>. This is code which can be pre-written and then shared with any component within any business domain. An example would be the generation and execution of SQL queries, or the generation of HTML pages using a templating engine.</li>
<li>
<b>Domain-specific</b>. This is code which needs to be written specifically for a particular business domain in order to execute the business rules required by that domain. The business layer will contain a number of different domain objects, and each object will contain it own subset of business logic.</li>
</ul>
<p>I have been building enterprise applications for nearly 4 decades, and my approach has always been to start building the components around the domain-agnostic logic and insert the domain-specific logic later. Why? Because in my experience the volume of domain-agnostic logic is far greater than the volume of domain-specific logic, and as there are large volumes of domain-agnostic logic which can be provided in reusable and sharable modules I have found it easier to build each user transaction by starting with the domain-agnostic logic and adding in the domain-specific logic later. Neither the front-end Presentation layer nor the back-end Data Access layer should contain any domain-specific logic, and the middle Business/Domain layer will always contain a mixture of standard domain-agnostic logic (the <b>invariant</b> parts of a Template Method) and non-standard domain-specific logic (the variant/variable methods in each subclass). With my technique I can build a raw user transaction from a template and add in the domain-specific logic afterwards at a faster rate than you can by starting with the domain-specific logic and adding in the domain-agnostic logic later.</p>
<p>In an enterprise application each user transaction has some sort of interaction with the database which, regardless of the table and the data it contains, follows a familiar and regular pattern. Each HTML document, while superficially different, has enough in common with other documents to enable it to be built from one of a small number of reusable <a href="https://www.tonymarston.net/php-mysql/infrastructure.html#xsl.stylesheet">XSL Stylesheets</a> which implement a series of reusable modules called templates. Each <a href="https://www.tonymarston.net/php-mysql/infrastructure.html#controller.script">Controller</a>, which translates user requests into actions on a particular <a href="https://www.tonymarston.net/php-mysql/infrastructure.html#table.class">Model</a>, can use the same pattern of actions on any number of different Models, so can be provided as a library of pre-written and reusable components.</p>
<p>The key to successful software development is the ability to spot these recurring templates and patterns, then to provide them with as much pre-written and reusable code as possible. This is the basis on which I developed my series of development frameworks. As far back as 1985 I built my first development framework in <a href="https://en.wikipedia.org/wiki/COBOL">COBOL</a> to provide as much domain-agnostic and sharable code as possible. In the 1990s I rewrote this in the UNIFACE language, and in 2003 I rewrote it again <a href="http://www.radicore.org">specifically for PHP</a>. With each rewrite the additional capabilities of each new language allowed me to build additional features into my development framework and also to provide more reusable components as well as shorter development times. My latest <a href="http://www.radicore.org">PHP framework</a> allows me to generate working user transactions which automatically contain all the necessary domain-agnostic code. The developer can then add in the necessary domain-specific code without having to touch the standard code which is provided by the framework. This is an easy job as the <a href="https://www.tonymarston.net/php-mysql/infrastructure.html#generic.table.class">abstract table class</a>, from which every <a href="https://www.tonymarston.net/php-mysql/infrastructure.html#table.class">concrete table class</a> inherits, contains a series of <a href="https://www.tonymarston.net/php-mysql/radicore-hooks.html">hook methods</a> specifically for this purpose.</p>
<p>In the introduction I identified the various components which are supposedly "required" when implementing <a href="https://en.wikipedia.org/wiki/Domain-driven_design">DDD</a>, but I have found a use for only some of them:</p>
<ul>
<li>
<b>Entities</b> - these are implemented as <a href="https://www.tonymarston.net/php-mysql/infrastructure.html#table.class">Model</a> classes with a separate class for each table in the database. Each of these classes inherits standard domain-agnostic code from an <a href="https://www.tonymarston.net/php-mysql/infrastructure.html#generic.table.class">abstract table class</a>. Entities are <i>stateful</i>.</li>
<li>
<b>Services</b> - these are pre-written and provided by the framework. They include <a href="https://www.tonymarston.net/php-mysql/infrastructure.html#view.object">Views</a> and <a href="https://www.tonymarston.net/php-mysql/infrastructure.html#controller.script">Controllers</a>. There is a separate View for each output format - HTML, PDF or CSV. There is a separate Controller for each <a href="https://www.tonymarston.net/php-mysql/dialog-types.html">Transaction Pattern</a>. Services are <i>stateless</i>.</li>
<li>
<b>Events</b> - an event is something that happens which is significant to those people who work in that domain. In a computer application it may also be known as a <b>unit of work</b>, <b>use case</b>, <b>user transaction</b> or <b>task</b>. It may mean updating the database to record the fact that something has happened in the outside world (such as a product has been sold to a customer), or it may mean interrogating the database to identify an action which needs to be performed in the outside world (such as generating a picklist so that items can be picked from inventory prior to being shipped to the customer). Regardless of the event's description to the user, each event/task can be boiled down to performing one or more <a href="https://en.wikipedia.org/wiki/Create,_read,_update_and_delete">CRUD</a> operations on one or more database tables. Each event/task is implemented as a user transaction which identifies a particular unit of work which will cause the database to be accessed in the appropriate manner. In the <a href="https://www.radicore.org">RADICORE</a> framework each has its own entry on the <a href="https://www.tonymarston.net/php-mysql/menuguide/section04.html#task">MNU_TASK</a> table in the <a href="https://www.tonymarston.net/php-mysql/menuguide/section04.html">framework database</a>. This data is then referenced in the <a href="https://www.tonymarston.net/php-mysql/menuguide/section04.html#menu">MNU_MENU</a> and <a href="https://www.tonymarston.net/php-mysql/menuguide/section04.html#navigation-button">MNU_NAV_BUTTON</a> tables to provide clickable links to activate each task. The <a href="https://www.tonymarston.net/php-mysql/menuguide/section04.html#role-task">ROLE_TASK</a> table is used to identify which tasks can be accessed by each <a href="https://www.tonymarston.net/php-mysql/menuguide/section04.html#role">USER_ROLE</a> which in turn controls the links which are visible to each user. The task then points to a <a href="https://www.tonymarston.net/php-mysql/infrastructure.html#component.script">component script</a> in the file system which identifies which action is to be performed on which object in the business/domain layer.</li>
<li>
<b>Repositories</b> - these are implemented as <a href="https://www.tonymarston.net/php-mysql/infrastructure.html#dml.class">Data Access Objects</a> with a separate DAO class for each supported DBMS. It is possible to switch an application from one DBMS to another simply by changing a single line in the configuration file.</li>
<li>
<b>Value objects</b> - PHP does not support them, so I don't use them.</li>
<li>
<b>Aggregates</b> - An aggregate is a collection of objects which are bound together by a root entity. These do not exist in any DBMS that I have used, so to put them in my software would introduce an artificial construct which would have no matching object in the database. Each table is a stand-alone object in the database, with its own structure, so in my software I create a stand-alone object in the domain/business layer to deal with that structure and apply its business rules. While each table has its own separate Model class, it is possible for a Controller to access more than one Model, and it is possible for one Model to access other Models if a single event/task needs to access multiple tables. Reading from multiple tables can be made more efficient by adding JOIN clauses to the SQL SELECT query.</li>
<li>
<b>Factories</b> - I have no need for any of these as each <a href="https://www.tonymarston.net/php-mysql/infrastructure.html#table.class">Model</a> class has only a single configuration, and that configuration, which involves turning an <a href="https://www.tonymarston.net/php-mysql/infrastructure.html#generic.table.class">abstract class</a> into a concrete class, is performed within the class constructor.</li>
</ul>
<p>For those of you who do not understand the difference between an <i>entity</i> and a <i>service</i>, an entity has <i>state</i> or data which can be loaded, changed and extracted over time. An example is a database record which has its current values extracted from the database, displayed in the UI where changes may be made and posted back, which causes the changed values to be updated in the database. A service on the other hand is <i>stateless</i> - data goes in, is processed in some way, then immediately spat out without hanging around for later extraction or manipulation. An example of a service is an <a href="https://www.tonymarston.net/php-mysql/infrastructure.html#xsl.transformation">XSL Transformation</a> which, after having transformed an XML document into an HTML web page, simply dies as there is nothing else left to do. The XML document is not retained for later use, it is discarded. Domain logic should only exist within an entity. A service should only contain logic which is required for the provision of that service and which is not specific to any domain.</p>
<p>Too many of today's programmers are taught to design their applications by starting with the software, which sits between the UI and the database, and leave the database till last as it is considered to be nothing more than "an implementation detail". In my early programming days I worked with badly structured software and badly designed databases, but I saw the light when I was taught that for a database application to be really successful the database must be properly designed by following the rules of <a href="https://www.tonymarston.net/php-mysql/database-design.html#normalisation">Data Normalisation</a> and, courtesy of a course on <a href="https://en.wikipedia.org/wiki/Jackson_structured_programming">Jackson Structured Programming</a>, that the software structure should be designed to follow the database structure as closely as possible. I personally witnessed the improvement that this combination of ideas made to both the initial software development and its subsequent maintenance, so I have absolutely no intention of turning my back on a winning formula. Just because someone has developed a different approach does not mean that I should follow it.</p>
<p>To me the reverse is true - the design of the database takes preference for the following reasons:</p>
<ul>
<li>It contains the vast majority of the business rules by identifying what data needs to be manipulated.</li>
<li>It must be designed according to the rules of <a href="https://www.tonymarston.net/php-mysql/database-design.html#normalisation">Data Normalisation</a> otherwise it may become difficult to build the user transactions which access that data.</li>
<li>I look upon each user transaction (<b>event</b> or <b>use case</b>) as a task which sits between the user interface (UI) and the database, and which incidentally executes sets of business rules. In this way I can build each task from a template and add in the business logic later instead of the less efficient approach of starting with the business logic and adding in the UI and database logic later.</li>
</ul>
<p>This approach falls into line with the following statement:</p>
<blockquote>Smart data structures and dumb code works a lot better than the other way around.
<p>Eric S. Raymond, "The Cathedral and the Bazaar"</p>
</blockquote>
<p>It does not matter that the data and business rules for each business domain are totally different when the common factor across all domains is that the data is stored in the database as tables and columns which are manipulated with SQL queries using just 4 basic operations - <a href="https://en.wikipedia.org/wiki/Create,_read,_update_and_delete">Create, Read Update and Delete</a>. Each table is treated as a separate object in the database, so I see no reason why each table should not have its own class/object in the application software. Each table contains certain business rules which are implied in its structure, so it is not unreasonable for the table's class to know its structure so that it can validate all input data against this structure. I particularly avoid the idea of <a href="https://en.wikipedia.org/wiki/Object_composition#Aggregation">object aggregation</a> where a single object in the software is responsible for multiple objects (tables) in the database. This, in my opinion, would be a violation of SRP. In the real world the concept of an ORDER may exist as a single entity, but when it appears in the database after normalisation it is split into separate ORDER_HEADER and ORDER_LINE objects. Each of these is a separate object with a separate structure and business rules, so each deserves its own separate table class in the software.</p>
<p>All the code which is common to every database table I put into an <a href="https://www.tonymarston.net/php-mysql/infrastructure.html#generic.table.class">abstract class</a>. Each table has its own name and collection of columns, so I specify these differences in the class constructor which inherits from the non-specific abstract class and turns it into a specific concrete class. While the collection of columns in a table's structure may be unique, their data types always come from a predefined list, so the validation of user input which checks that the value for a column matches that column's data type can also be predefined. Each table has its own primary key, optional unique keys and relationships with other tables, so these can also be identified in the class constructor and handled by code which is inherited from the abstract class. <a href="https://www.tonymarston.net/php-mysql/infrastructure-faq.html#faq40">Primary data validation</a> is carried out by a standard <a href="https://www.tonymarston.net/php-mysql/infrastructure.html#validation.class">validation class</a> which compares all input data with that table's <a href="https://www.tonymarston.net/php-mysql/data-dictionary.html#%24fieldspec">field specifications</a>. This means that the developer does not have to write any code to perform this task, it is automatically handled by the framework.</p>
<p>The addition of specific business rules for particular tables is handled using the <a href="https://www.tonymarston.net/php-mysql/template-method-pattern-as-a-framework.html#characteristics">Template Method Pattern</a> whereby the <a href="https://www.tonymarston.net/php-mysql/infrastructure.html#generic.table.class">abstract table class</a> contains a series of <a href="https://www.tonymarston.net/php-mysql/radicore-hooks.html">"hook" methods</a> which are empty by default. If a developer wishes to add custom code into a particular step of the processing cycle all he has to do is copy the relevant empty "hook" method from the abstract class into his concrete class, then insert the necessary code into this method.</p>
<p>This means that to deal with a new domain with a new database all I need do is follow these steps:</p>
<ol>
<li>Build the database.</li>
<li>Build the <a href="https://www.tonymarston.net/php-mysql/infrastructure.html#model">Model class</a> for each table in that database. These are generated by the framework.</li>
<li>Build the user transactions for each table. These are generated by the framework and provide basic functionality using Template Methods which have been built into the <a href="https://www.tonymarston.net/php-mysql/infrastructure.html#generic.table.class">abstract table class</a>
</li>
<li>Add in table-specific logic by inserting code into the relevant <a href="https://www.tonymarston.net/php-mysql/radicore-hooks.html">"hook" method</a> in that <a href="https://www.tonymarston.net/php-mysql/infrastructure.html#table.class">table subclass</a>
</li>
</ol>
<p>I do not need to build any <a href="https://www.tonymarston.net/php-mysql/infrastructure.html#controller">Controllers</a> or <a href="https://www.tonymarston.net/php-mysql/infrastructure.html#view">Views</a> as these are provided by the framework as pre-built and reusable components. Building user transactions is as simple as selecting a database table then joining it with one of the <a href="https://www.tonymarston.net/php-mysql/dialog-types.html">Transaction Patterns</a> which is also built into the framework. This is explained in <a href="https://www.tonymarston.net/php-mysql/dont-do-domain-driven-design.html#use.the.framework">Use the framework to build components</a>.</p>
<p>When I started my programming career with <a href="https://en.wikipedia.org/wiki/COBOL">COBOL</a> in the 1970s a common complaint was that the language was so verbose with its numerous divisions (identification division, environment division, data division and procedure division) that it took a long time to write even a simple program. Then a person much older and wiser than me pointed out a simple fact - a good programmer need only ever write <i>one</i> program from scratch, after which he could copy this code and amend it as necessary to create a different program. As COBOL was a language designed for business applications, and all these applications used some sort of database, there was a great deal of code which moved the data from the database to the screen and from the screen to the database by constructing and executing queries. Note that SQL did not exist at that time, so a "query" was actually a hard-coded function call that used the non-relational database that was provided with the operating system. The differences between one program and the next were then limited to the name of the database table and the names of its columns, plus the business rules which were unique to that table. My goal then became to write code in such a way that it was easy to identify that which was similar from that which was different so that I could minimise the changes necessary to make the code work with a different table and its different set of columns. Note that I use the word "similar" and not "identical". Each program may have a block of code which moves data from the database to the screen, and another block which moves data from the screen to the database, but each block contains a series of <code>MOVE source TO destination</code> statements where the "source" and "destination" names were hard-coded and could not be parameterised.</p>
<p>All I could do initially was to put identical code into a shared library so that the library function could be called instead of the code within that function being duplicated. I also began to write library functions which dealt with "similar" conditions by passing the differences as arguments so that the function body could be kept as static as possible. The results of this work is identified in <a href="https://www.tonymarston.net/cobol/standardutilities.html">Library of Standard Utilities</a> on my <a href="https://www.tonymarston.net/cobol/index.html">COBOL</a> page.</p>
<p>As explained in <a href="https://www.tonymarston.net/php-mysql/what-is-a-framework.html">What is a Framework?</a> you have to write your own code in order to call a library routine, whereas a framework contains components which call your code. A framework should also allow you to create application components by writing less code.</p>
<p>I converted my library into a framework while working on a project in 1985. Up until that point the practice had been to hard-code each of the menu screens so that the user could see a list of options and choose one to be executed. A menu had to fit within the screen's dimensions (no scrolling was allowed), so each menu page had to be written as a separate subprogram. Note the difference between a "program" and a "subprogram" - a program can only be activated from the command line, while a subprogram can only be called from within a program or subprogram. The starting point for each application was always a logon screen which accepted a user ID and password, and from then on it passed control to the top-level menu where it waited for the user to choose which option to run next. Each application therefore had its own copy of the logon screen, its own set of hard-coded menu screens, plus its own set of application components. Not all users had access to all components, so there was a primitive mechanism which identified who could access what. Unfortunately this condition could only be evaluated <i>after</i> the component was loaded, so while all users could see all options on the menu screen, they sometimes selected an option only to be told "You don't have access to this option".</p>
<p>This all changed when the customer made a simple request: "If a user does not have access to an option, then that option should not appear on the menu screen". This meant that I had to change from using <i>static</i> menus to a system of <i>dynamic</i> menus. The request was made on a Friday, so I spent a couple of hours over the weekend in designing a solution and started implementing it on the Monday. By the following Friday the implementation was complete. The solution involved the creation of a new set of database tables plus the programs to maintain them. These are documented in the <a href="https://www.tonymarston.net/cobol/menu-user.html">Menu and Security System User Manual</a> as follows:</p>
<ul>
<li>The <a href="https://www.tonymarston.net/cobol/menu-user.html#mgr04">M-TRAN</a> table identified every option available in the application.</li>
<li>The <a href="https://www.tonymarston.net/cobol/menu-user.html#mgr05">M-USER</a> table identified every person who was allowed to access the application.</li>
<li>The <a href="https://www.tonymarston.net/cobol/menu-user.html#mgr07">D-TRAN-USER</a> table identified which options each user was allowed to access.</li>
<li>The <a href="https://www.tonymarston.net/cobol/menu-user.html#mgr06">D-MENU</a> table identified which option was available on which menu screen.</li>
</ul>
<p>This arrangement allowed menu screens to be built or modified by changing the contents of a few database tables without having to change any program code. When a menu option was selected the contents of the D-MENU table was read, and any options which were not on the D-TRAN-USER table for that user were filtered out. This meant that the menu screen never displayed an option which the user could not select. This also removed the need to have any security checking within any subprograms as such checking was now performed by the framework <i>before</i> the subprogram was called instead of within the subprogram <i>after</i> it was called. This software became a framework by virtue of the fact that it provided a standard set of pre-written code which could be reused in any application without having to be redesigned and rewritten each time. The programmer no longer had to write any code to call his application components as the standard code did this all for him. This meant that the programmer could concentrate on writing the application components and not waste time on the code which called those components.</p>
<p>This database structure was carried forward when I redeveloped the framework in each new language, first UNIFACE in the 1990s and then PHP in 2003. Each of these languages came with its own set of new facilities which either allowed more options to be provided, or allowed more code to be transferred to reusable functions. The results of the rewrite in PHP have proved to be very successful as I am now able to build application components at a much faster rate and with more features built in as standard than I could with its previous incarnations.</p>
<p>The challenge with PHP was that it was the first language I had used which had Object Oriented (OO) capabilities, so I had to learn how to use these things called <a href="https://www.tonymarston.net/php-mysql/oop-for-heretics.html#encapsulation">Encapsulation</a>, <a href="https://www.tonymarston.net/php-mysql/oop-for-heretics.html#inheritance">Inheritance</a> and <a href="https://www.tonymarston.net/php-mysql/oop-for-heretics.html#polymorphism">Polymorphism</a> to maximum advantage. According to some people my implementation of these principles is <a href="https://www.tonymarston.net/php-mysql/your-code-is-crap.html">totally wrong</a>, but their arguments are futile when you consider that my results are superior to theirs.</p>
<p>The steps I took to build the components to maintain the first database table in my framework are detailed in <a href="https://www.tonymarston.net/php-mysql/anatomy-of-an-enterprise-application.html#building.components">Building the first components</a>. This resulted in the creation of 1 Model class and 6 Controllers.</p>
<p>I then copied these 7 scripts and renamed them to perform the same actions on a second database table, as described in <a href="https://www.tonymarston.net/php-mysql/anatomy-of-an-enterprise-application.html#duplicating.components">Duplicating and Refactoring components</a>. As this resulted in a huge amount of duplicated code I then went through a refactoring exercise in order to put as much common code as possible into reusable modules. For the two Model classes I created an <a href="https://www.tonymarston.net/php-mysql/infrastructure.html#generic.table.class">abstract table class</a> for the common code which could be inherited, then moved code from the table class into the abstract class. When I was finished I had nothing left in each table class except the constructor which did nothing but supply the table name and its structure.</p>
<p>Then I turned my attention to the Controllers. Instead of having a separate Controller which contained hard-coded references to a particular Model class I found a way to make them work with a class name that was not supplied until run-time. This mechanism is called a <a href="https://www.tonymarston.net/php-mysql/infrastructure.html#component.script">component script</a>, and there is one of these for every user transaction in the application. If you study my implementation carefully you should see that each Controller is dependent on a Model (sometimes two or three models), but rather than this dependency being hard-coded it is in fact injected at run-time, thus demonstrating a form of <a href="https://www.tonymarston.net/php-mysql/dependency-injection-is-evil.html#inject.model.into.the.controller">Dependency Injection</a>. Because the methods called by each Controller are defined within the <a href="https://www.tonymarston.net/php-mysql/infrastructure.html#generic.table.class">abstract table class</a> which is inherited by every <a href="https://www.tonymarston.net/php-mysql/infrastructure.html#table.class">concrete table class</a> this means that each Controller is capable of calling the same methods on each of the 400+ table classes in my enterprise application, thus demonstrating one of the goals of OOP which is called <a href="https://www.tonymarston.net/php-mysql/oop-for-heretics.html#polymorphism">Polymorphism</a>.</p>
<p>Over time I discovered that some user transactions were becoming more complicated than could be covered by my original 6 Controllers, so I gradually added to their number until I now have over 50 of them which are documented in <a href="https://www.tonymarston.net/php-mysql/dialog-types.html">Transaction Patterns for Web Applications</a></p>
<p>Initially I had to generate the scripts for new table classes and user transactions by hand, but as all these followed a standard pattern I decided to automate them as much as possible. To handle the class files I followed the steps described in <a href="https://www.tonymarston.net/php-mysql/anatomy-of-an-enterprise-application.html#class.generation">Generating class files</a>. To handle the user transactions I followed the steps described in <a href="https://www.tonymarston.net/php-mysql/anatomy-of-an-enterprise-application.html#task.generation">Generating user transactions</a>. This then meant that I could create a new table in my database, then generate <i>and run</i> the 6 basic user transactions in 5 minutes without writing a single line of code - no PHP, no HTML, no SQL. All the domain-agnostic code is provided by the framework while it is only the domain-specific code which needs to be written by the developer. This is why my productivity is higher than my rivals both for new development and subsequent maintenance.</p>
<p>Creating a new subsystem/domain follows the same basic steps:</p>
<ul>
<li>The analysis and design phase identifies the database design, the tasks and the business rules.</li>
<li>Design and build the database.</li>
<li>Import the database structure into my <a href="https://www.tonymarston.net/php-mysql/data-dictionary.html">Data Dictionary</a>.</li>
<li>Export each table from the Data Dictionary to produce a separate Model class for each table.</li>
<li>Select a table, link it with a <a href="https://www.tonymarston.net/php-mysql/dialog-types.html">Transaction Pattern</a>, then press a button to generate a working transaction. Repeat using as many different transaction patterns as required.</li>
<li>Modify table class files to add any business rules or task-specific behaviour. This can be done by using the <a href="https://www.tonymarston.net/php-mysql/radicore-hooks.html">hook methods</a> which are already provided by the framework. These are empty methods which are waiting to be filled with custom code.</li>
</ul>
<p>Note here that because so much work is provided by pre-built and reusable framework components the bulk of the developer's time can be spent in dealing with the most important part of the software in any application - the business rules. The developer doesn't have to design any software around those business rules, the framework creates the class files for each database table and the developer adds code into the relevant <a href="https://www.tonymarston.net/php-mysql/radicore-hooks.html">hook method</a> in each class file to process the business rules.</p>
<p>Time after time I see OO tutorials which say that after a use case has been identified a class should be constructed with a method name specifically to execute that particular use case. I do not. Instead of using method names which are unique to the operation within each domain I use names which are common to the underlying database activity that will be performed during the execution of that use case. All the method names that can be used by my <a href="https://www.tonymarston.net/php-mysql/infrastructure.html#controller.script">Controllers</a> when communicating with <a href="https://www.tonymarston.net/php-mysql/infrastructure.html#table.class">Models</a> have already been predefined as public methods in my <a href="https://www.tonymarston.net/php-mysql/infrastructure.html#generic.table.class">abstract table class</a> which is inherited by every <a href="https://www.tonymarston.net/php-mysql/infrastructure.html#table.class">concrete table class</a>. This simple practice allows any of my Controllers to be used with any of my Models. This is a prime example of <a href="https://www.tonymarston.net/php-mysql/oop-for-heretics.html#loose.coupling">loose coupling</a>, which is supposed to be good. If I were to manually create a method in a Model to execute a particular use case then I would have to manually create a Controller to call that method on that Model. As that method would be unique to that Model then that Controller could be used only with that Model, thus making that Controller inextricably tied to that Model and thus not reusable with any other Model. This is a prime example of <a href="https://www.tonymarston.net/php-mysql/oop-for-heretics.html#tight.coupling">tight coupling</a>, which is supposed to be bad.</p>
<p>Instead I create a task in the framework's database for each and every use case. Each task points to a <a href="https://www.tonymarston.net/php-mysql/infrastructure.html#component.script">component script</a> in the file system which is very small as all it does is identify which Model should be used with which View and which Controller. This means that instead of custom method names I use generic names, such as:</p>
<table border="1" cellspacing="0" cellpadding="0">
<colgroup>
<col>
<col>
<col>
</colgroup>
<tr>
<th>OO purists</th>
<th>effect on the database</th>
<th>the Tony Marston way</th>
</tr>
<tr>
<td>createProduct()</td>
<td>insert a record into the PRODUCT table</td>
<td>
<pre>
<span>$table_id</span> = <span>'product'</span>;
....
require <span>"classes/<span>$table_id</span>.class.inc"</span>;
<span>$dbobject</span> = new <span>$table_id</span>;
<span>$result</span> = <span>$dbobject</span>-&gt;insertRecord(<span>$_POST</span>);
</pre>
</td>
</tr>
<tr>
<td>createCustomer()</td>
<td>insert a record into the CUSTOMER table</td>
<td>
<pre>
<span>$table_id</span> = <span>'customer'</span>;
....
require <span>"classes/<span>$table_id</span>.class.inc"</span>;
<span>$dbobject</span> = new <span>$table_id</span>;
<span>$result</span> = <span>$dbobject</span>-&gt;insertRecord(<span>$_POST</span>);
</pre>
</td>
</tr>
<tr>
<td>createInvoice()</td>
<td>insert a record into the INVOICE table</td>
<td>
<pre>
<span>$table_id</span> = <span>'invoice'</span>;
....
require <span>"classes/<span>$table_id</span>.class.inc"</span>;
<span>$dbobject</span> = new <span>$table_id</span>;
<span>$result</span> = <span>$dbobject</span>-&gt;insertRecord(<span>$_POST</span>);
</pre>
</td>
</tr>
<tr>
<td>payInvoice()</td>
<td>insert a record into the PAYMENT table</td>
<td>
<pre>
<span>$table_id</span> = <span>'payment'</span>;
....
require <span>"classes/<span>$table_id</span>.class.inc"</span>;
<span>$dbobject</span> = new <span>$table_id</span>;
<span>$result</span> = <span>$dbobject</span>-&gt;insertRecord(<span>$_POST</span>);
</pre>
</td>
</tr>
</table>
<p>A significant point to notice here is that I do NOT have a separate Controller for each Model which calls methods which are specific to that Model as this would require each Controller and Model to be hand-crafted and tightly coupled to each other. All my Controllers are supplied by the framework as services, and each Model is generated by the framework as an entity. All my Controllers access their Model(s) by using method names which have been defined in my <a href="https://www.tonymarston.net/php-mysql/infrastructure.html#generic.table.class">abstract table class</a>, and each Model inherits from this same <a href="https://www.tonymarston.net/php-mysql/infrastructure.html#generic.table.class">abstract table class</a>. This means that no Controller is inextricably tied to a single Model, which would produce <a href="https://www.tonymarston.net/php-mysql/oop-for-heretics.html#tight.coupling">tight coupling</a>, but instead any Controller can be used with any Model, thus becoming as <a href="https://www.tonymarston.net/php-mysql/oop-for-heretics.html#loose.coupling">loosely coupled</a> as is possible.</p>
<p>Anybody who knows anything about OOP should see that my method is a shining example of <a href="https://www.tonymarston.net/php-mysql/oop-for-heretics.html#polymorphism">polymorphism</a> in action. Every one of my table classes uses the same method signatures by virtue of the fact that they are provided by the same <a href="https://www.tonymarston.net/php-mysql/infrastructure.html#generic.table.class">abstract table class</a>. Every one of my <a href="https://www.tonymarston.net/php-mysql/infrastructure.html#controller.script">page controllers</a> calls these methods, so if I have 50 controllers and 400 table classes that means that I have 50 x 400 = 20,000 - yes TWENTY THOUSAND - opportunities for polymorphism. How many frameworks have you seen which can match that?</p>
<p>If you show the average programmer a collection of different HTML pages they will tell you that each is totally unique and therefore needs to be designed and built independently. An experienced programmer should be able to see the similarities as well as the differences. The similarities are elements which appear in multiple pages and which look and act in the same way, and it should be possible to create a catalog of patterns which show the different arrangements of these similar elements. Among these similar and repeating elements are the following:</p>
<ul>
<li>The <a href="https://www.tonymarston.net/php-mysql/dialog-types.html#menu-bar">Menu Bar</a> which contains the current set of menu options.</li>
<li>The <a href="https://www.tonymarston.net/php-mysql/dialog-types.html#title-bar">Title Bar</a> which displays the name of the current task.</li>
<li>The <a href="https://www.tonymarston.net/php-mysql/dialog-types.html#navigation-bar">Navigation Bar</a> which provides a number of menu options which are only applicable to the current task.</li>
<li>A line of <a href="https://www.tonymarston.net/php-mysql/dialog-types.html#column-headings">Column Headings</a> which appear above rows of data in LIST screens.</li>
<li>A <a href="https://www.tonymarston.net/php-mysql/dialog-types.html#data-area">Data area</a> which contains application-specific data.</li>
<li>A <a href="https://www.tonymarston.net/php-mysql/dialog-types.html#message-area">Message Area</a> which shows any error messages in red and informative messages in green.</li>
<li>A <a href="https://www.tonymarston.net/php-mysql/dialog-types.html#pagination">Pagination Area</a> which appears under those areas in LIST screens which contain multiple rows displayed horizontally.</li>
<li>A <a href="https://www.tonymarston.net/php-mysql/dialog-types.html#scrolling">Scrolling Area</a> which appears under those areas which contain a single row displayed vertically.</li>
<li>The <a href="https://www.tonymarston.net/php-mysql/dialog-types.html#action-bar">Action Bar</a> which identifies actions such as SUBMIT or CANCEL.</li>
</ul>
<p>Long ago, after having designed thousands of different forms, I recognised that each screen/form could be described in the following ways:</p>
<ul>
<li>
<b>Structure</b> - what it looks like, how many elements it contains, how they are positioned and how they are related.</li>
<li>
<b>Behaviour</b> - what it does, what actions it performs, either automatically or by user selection.</li>
<li>
<b>Content</b> - what data from which database table(s) is visible within each element.</li>
</ul>
<p>Unlike with my previous languages which required the forms to be built and compiled using a built-in process, the building of HTML pages is far easier simply because each page is constructed as a text file which does not have to be compiled and therefore can be constructed <a href="https://en.wikipedia.org/wiki/On_the_fly">on the fly</a>. This text file can be constructed in many ways, but my method of choice was to put all the relevant data into an <a href="https://www.tonymarston.net/php-mysql/infrastructure.html#xml.file">XML document</a> and then transform it into <a href="https://www.tonymarston.net/php-mysql/infrastructure.html#html.output">HTML</a> using an <a href="https://www.tonymarston.net/php-mysql/infrastructure.html#xsl.stylesheet">XSL stylesheet</a>. I started off with a separate XSL stylesheet for each HTML page, but following a process of refactoring I managed to reduce this down to a small set of <a href="https://www.tonymarston.net/xml-xsl/reusable-xsl.html">Reusable XSL Stylesheets</a>. I created a single reusable <a href="https://www.tonymarston.net/php-mysql/infrastructure.html#view.object">View</a> object to handle both the construction of the XML document and the <a href="https://www.tonymarston.net/php-mysql/infrastructure.html#xsl.transformation">XSL transformation</a>, which then meant that I did not have to write any code, apart from the small <a href="https://www.tonymarston.net/php-mysql/infrastructure.html#screen.structure">screen structure script</a>, in order to create any web page.</p>
<p>Some pages produce <a href="https://www.tonymarston.net/php-mysql/output-to-pdf.html">PDF output</a> instead of HTML, so again I have a single View object which handles all the processing with the aid of a small <a href="https://www.tonymarston.net/php-mysql/output-to-pdf.html#report.structure.file">report structure file</a>.</p>
<p>Some of the business rules are implemented in the structure of the database which identifies what data elements need to be stored for each entity. Each of these elements has a specific type and size, and can be designated as being either NULL (optional) or NOT NULL (required). A competent programmer should therefore be able to create a standard routine which validates that each piece of user input matches the database definition before any attempt is made to add it to he database. In my framework this functionality is provided by my <a href="https://www.tonymarston.net/php-mysql/infrastructure.html#validation.class">validation class</a>, so it is an example of more code which does not have to be written by the developer.</p>
<p>Each method call from the Controller to the Model will initiate a database operation in a series of predefined steps which have been built into the <a href="https://www.tonymarston.net/php-mysql/infrastructure.html#generic.table.class">abstract table class</a> as shown in this set of <a href="https://www.tonymarston.net/php-mysql/uml-diagrams.html">UML diagrams</a>. Those methods with a "_cm_" prefix, which are known as <a href="https://www.tonymarston.net/php-mysql/functions-and-variables.html#customisable.methods">customisable methods</a> or "hook" methods, are defined within the abstract class but are empty, which means that when they are executed nothing happens. However, if any of these methods is copied from the abstract class to the concrete class then the copy in the concrete class will override, or be executed instead of, the empty original in the abstract class. If any code is inserted into the method in the concrete class then this code will be executed when that method is called.</p>
<p>The use of these "hook" methods is a prime example of the <a href="https://www.tonymarston.net/php-mysql/radicore-hooks.html#template.method">Template Method pattern</a>.</p>
<p>It is quite possible that it some circumstances you may wish to perform some processing in a particular hook method for a particular task that is totally different from the code used in the same method for a different task. In this situation I use a standard OO technique called <a href="https://en.wikipedia.org/wiki/Inheritance_(object-oriented_programming)#Subclasses_and_superclasses">subclassing</a>. I create a subclass of the table class with a name such as <code>&lt;table&gt;_sNN</code> where "s" denotes a subclass and "NN" is an incrementing number. This automatically inherits all the methods in the superclass but allows me to override any of those methods with a different implementation. Note that the subclass deals with the same table as the superclass, it just provides different business rules for that table. I then change that task's <a href="https://www.tonymarston.net/php-mysql/infrastructure.html#component.script">component script</a> to point to this subclass. You can see working examples of this in the Data Dictionary subsystem of the framework where you will find the following class files:</p>
<ul>
<li>
<code>dict_table.class.inc</code> handles the standard processing for "dict_table".</li>
<li>
<code>dict_table_s01.class.inc</code> is used to import table details from the database schema into the dictionary database.</li>
<li>
<code>dict_table_s02.class.inc</code> is used to export table details from the dictionary database into the PHP application.</li>
</ul>
<p>When I say that I don't do Domain Driven Design I mean that I do not go through a separate design process for the software of each domain. Each subsystem/domain in my ERP application has one important characteristic which is shared with all the other subsystems/domains - they are all database applications. Regardless of what data a table contains every table has similar characteristics in that the only operations which can be performed on them are the <a href="https://en.wikipedia.org/wiki/Create,_read,_update_and_delete">CRUD</a> operations. Every event/task, regardless of what it does for the user, can be boiled down to performing one or more CRUD operations on one or more tables, then displaying the result as either HTML, CSV or PDF. After building thousands of events I have spotted numerous recurring patterns of structure and behaviour which are now encapsulated in my library of <a href="https://www.tonymarston.net/php-mysql/dialog-types.html">Transaction Patterns</a>.</p>
<p>Once I have designed the database for a new domain I do not have to go through a separate exercise to design the software for that domain as I can use my <a href="https://www.tonymarston.net/php-mysql/data-dictionary.html">Data Dictionary</a> to create a separate <a href="https://www.tonymarston.net/php-mysql/infrastructure.html#table.class">Model</a> class for each table in the database. When I identify an event/task which needs to be performed on a table I can create a basic but functional task simply by joining a <a href="https://www.tonymarston.net/php-mysql/dialog-types.html">Transaction Pattern</a> to that database table. In this way I can use my <a href="https://www.tonymarston.net/php-mysql/infrastructure.html">framework</a> to both build <i>and run</i> database applications. Not just a database application for a particular domain, but <i>any</i> database application for <i>any</i> domain. This framework allows me to generate the classes for every domain object to fit a standard pattern, and then to generate the user transactions for that domain object to fit a standard pattern. You could say that my framework is an example of the <a href="https://www.tonymarston.net/php-mysql/template-method-pattern-as-a-framework.html">Template Method Pattern</a> on steroids.</p>
<p>An enterprise application is comprised of a series of subsystems or modules each of which serves a different business area or domain. Each subsystem contains a series of user transactions which perform an action or unit of work for the user. Each transaction has a user interface at the front end, a database at the back end, an a component in the middle which moves data between the two ends as well as executing the business rules which are specific to that domain and/or user transaction. Of the two types of logic - domain-specific and domain-agnostic - it should become apparent to the experienced developer that the domain-agnostic logic is the larger of the two, and contains elements which are common to many transactions regardless of their business domain and which could be placed into reusable components so that they can be shared instead of duplicated. If every user transaction follows a familiar pattern with the only difference being the domain-specific logic then it would seem a good idea, at least it would to me, to start building each user transaction from a pre-defined pattern into which the domain-specific logic can be added afterwards.</p>
<p>This situation appears to me as being a prime candidate for the <a href="https://www.tonymarston.net/php-mysql/radicore-hooks.html#template.method">Template Method pattern</a>. In my implementation of this design pattern my framework provides a series of <a href="https://www.tonymarston.net/php-mysql/dialog-types.html">Transaction Patterns</a> each of which handles a different combination of database interaction and user interface. In the enterprise application which I built using my framework I have <a href="https://www.tonymarston.net/php-mysql/dont-do-domain-driven-design.html#multiple.domains">multiple subsystems/domains</a> which contain over 3,000 user transactions which were built from 50 patterns. While the sharable Controllers contain fixed logic each Model contains a combination of fixed logic, which is defined in and inherited from the abstract class, and customisable logic which can be defined in the various "hook" methods within each concrete class. These hook methods are initially defined as empty methods in the abstract class, but which can be copied into the concrete class and filled with suitable code.</p>
<p>This means that I do not have to waste valuable time in designing classes and methods for each object in the business layer as each object is a database table whose basic operations can be provided by methods in the <a href="https://www.tonymarston.net/php-mysql/infrastructure.html#generic.table.class">abstract table class</a>. Each table therefore has its own concrete class which I can generate, by means of my Data Dictionary, from the database schema. I do not have any class hierarchies as every concrete table class (Model) inherits all the standard code it requires from my <a href="https://www.tonymarston.net/php-mysql/infrastructure.html#generic.table.class">abstract table class</a>. Domain-specific logic can be added into a Model class using any of the numerous <a href="https://www.tonymarston.net/php-mysql/radicore-hooks.html">hook methods</a>. As each user transaction has its own combination of user interface and interaction with the database I have managed to create a library of 50 <a href="https://www.tonymarston.net/php-mysql/dialog-types.html">Transaction Patterns</a> each of which covers a different combination. Unlike design patterns which exist only as abstract descriptions my Transaction Patterns exist as concrete implementations which can be reused over and over again. In my framework it is possible to generate and then run a working user transaction simply by saying "Take table X, link it with pattern Y and produce transaction Z" thus eliminating the need to write any of that boring boilerplate code.</p>
<p>People should remember that the primary objective of a software developer is to develop cost-effective software for the benefit of the paying customer. It is *NOT* to follow a set of arbitrary or artificial rules created by fellow developers to promote their idea of "purity" or "perfection". I have encountered many of these rules in the past, and I have found that I can write better software by ignoring them. The only universal rule that I follow comes from a book called "The Structure and Interpretation of Computer Programs" which was published by H. Abelson and G. Sussman in 1984:</p>
<blockquote>Programs must be written for people to read, and only incidentally for machines to execute.</blockquote>
<p>Other rules that I follow are "<a href="https://en.wiktionary.org/wiki/KISS_principle">Keep It Simple, Stupid</a>" and "<a href="https://en.wiktionary.org/wiki/if_it_ain%27t_broke,_don%27t_fix_it">If it ain't broke don't fix it</a>".</p>
<p>In OOP one of the most important principles is that of <b>abstraction</b> from which we can build abstract classes to identify the high-level concepts, and concrete classes which supply the low-level details. It could be said that an abstract class provides a basic pattern while a concrete class provides a working implementation of that pattern. In order to fulfil one of the aims of OOP, which is to provide more reusable code, it should be noted that creating a large abstract class which is inherited hundreds of times is far better than creating a small abstract class which is only inherited once or twice.</p>
<p>If an enterprise application is comprised of a mixture of domain-specific logic and domain-agnostic logic, and the domain-agnostic logic can be seen to contain a set of repeating patterns of behaviour, it would make sense to me to create some sort of abstraction for each pattern so that it can be converted into any number of concrete implementations when building the individual application components. It should then be obvious that the domain-specific logic resides in the concrete implementation and not the abstract pattern. Since the abstraction should be defined BEFORE the concrete implementation then the followers of DDD have got it arse-backwards by starting with the domain-specific logic and leaving the domain-agnostic logic until last. This to me sounds as daft as building a house by starting with the roof and leaving the foundations until last.</p>
<p>When I add another domain to my enterprise application it is implemented as another subsystem which has its own subdirectory under the application's root directory, has its own database to hold its collection of tables, and has its own set of events/tasks in the MENU database. After analysing the requirements of the new subsystem I first build the database, construct the class file for each database table, then use the framework to build the tasks which perform operations on each table class. Specific business logic is added in afterwards by inserting code into any of the pre-defined <a href="https://www.tonymarston.net/php-mysql/radicore-hooks.html">hook methods</a>.</p>
<a name="comparison.matrix" id="comparison.matrix"></a>
<p>The differences between the "approved" DDD approach and my heretical approach can be summarised in the following comparison matrix:</p>
<table border="1" cellspacing="0" cellpadding="0">
<colgroup>
<col>
<col>
<col>
</colgroup>
<tr>
<th></th>
<th>the "approved" way</th>
<th>the Tony Marston way</th>
</tr>
<tr>
<td>1</td>
<td>Start by focusing on the domain logic and design the software around that logic.</td>
<td>No. After analysing the requirements I have a preliminary database design and a list of events/tasks (use cases). I then build the database and design the software based on actions that will be performed on those tables.</td>
</tr>
<tr>
<td>2</td>
<td>Design classes which have a unique method/operation for each use case.</td>
<td>No. If you have a large number of unique method names then you have fewer opportunities for code reuse via <a href="https://www.tonymarston.net/php-mysql/what-is-oop.html#polymorphism">polymorphism</a>, so the more method names which are identical the better. Each event has a separate task in the MENU database which points to a separate <a href="https://www.tonymarston.net/php-mysql/infrastructure.html#component.script">component script</a> which in turn identifies a Model, View and Controller which will be used to implement that task. Every Controller communicates with its Model(s) using standard method names which every Model inherits from the same <a href="https://www.tonymarston.net/php-mysql/infrastructure.html#generic.table.class">abstract table class</a>.</td>
</tr>
<tr>
<td>3</td>
<td>Design complex class hierarchies.</td>
<td>No. I never inherit from one concrete class to create another concrete class. It is common to use the <a href="https://en.wikipedia.org/wiki/Is-a">IS-A</a> test to identify a relationship between two entities, and in my architecture each Model "is-a" database table so each concrete table class inherits from a single <a href="https://www.tonymarston.net/php-mysql/infrastructure.html#generic.table.class">abstract table class</a>.</td>
</tr>
<tr>
<td>4</td>
<td>Use object aggregation to create objects which are responsible for groups of database tables.</td>
<td>No. The idea of a "root" table through which you have to go in order to access individual tables within that group simply does not exist in any RDBMS that I have used. Each table is a stand-alone entity, so I give each table a stand-alone class.</td>
</tr>
<tr>
<td>5</td>
<td>Use object composition instead of inheritance.</td>
<td>No. There is nothing wrong with inheritance <i>when it is used properly</i>, but it would appear that far too many of today's young programmers simply do not know what "properly" means. According to the book <a href="https://en.wikipedia.org/wiki/Design_Patterns">Design Patterns: Elements of Reusable Object-Oriented Software</a> problems can be created when you have deep inheritance hierarchies (which I don't) and extend a concrete class to create a totally difference concrete class (which I don't). You can avoid these problems by only ever inheriting from abstract classes (which I do). I have a single <a href="https://www.tonymarston.net/php-mysql/infrastructure.html#generic.table.class">abstract table class</a> which is inherited by every <a href="https://www.tonymarston.net/php-mysql/infrastructure.html#table.class">concrete table class</a>.</td>
</tr>
<tr>
<td>6</td>
<td>Create objects to deal with domain events.</td>
<td>No. A domain event is simply a user transaction which performs one or more operations on one or more database tables with the addition of any business rules. Objects in the Business/Domain layer are centered around entities (database tables) which have properties (data) and methods (operations). These objects represent nouns while the operations represent verbs. Objects in the Presentation or Data Access layers which perform standard operations on the data obtained from business objects are called Services. Domain objects have state while Services do not. Domain objects are large in number and are specific to their individual domains. Services are far smaller in number, are domain-agnostic, and are provided by the framework. Domain Events are therefore implemented as operations within the relevant domain object and not as separate service objects.</td>
</tr>
<tr>
<td>7</td>
<td>Create service objects to deal with operations which conceptually do not belong to any object.</td>
<td>No. All such objects have already been built and come supplied with the framework. These are <a href="https://www.tonymarston.net/php-mysql/infrastructure.html#controller.script">Controllers</a>, <a href="https://www.tonymarston.net/php-mysql/infrastructure.html#view.object">Views</a> and <a href="https://www.tonymarston.net/php-mysql/infrastructure.html#dml.class">Data Access Objects</a>.</td>
</tr>
<tr>
<td>8</td>
<td>Use factories to create domain objects.</td>
<td>No. I don't need factories to configure any domain object as each object has only a single implementation, and that is handled with the class constructor.</td>
</tr>
<tr>
<td>9</td>
<td>Create a separate repository for each domain object.</td>
<td>No. I have a separate DAO for each RDBMS which can deal with any table with that RDBMS. I do not have a separate DAO for each table (domain object). Every table in the database is subject to the same operations - <a href="https://en.wikipedia.org/wiki/Create,_read,_update_and_delete">Create, Read, Update and Delete</a> - which are specified using a query string, so a competent programmer should have no difficulty in building a single object which can handle those four operations on any database table.</td>
</tr>
<tr>
<td>10</td>
<td>Create a separate Controller for each Model.</td>
<td>No. Each user transaction performs one or more operations on one or more database tables, and as the operations which can be performed on any database table are exactly the same it is possible for a Controller to perform these operations on a table whose identity is not supplied until runtime. Each Model (table class) inherits the same set of public methods from a single <a href="https://www.tonymarston.net/php-mysql/infrastructure.html#generic.table.class">abstract table class</a>, so a Controller can be used with any Model which implements these methods - which in my case is ALL of them.</td>
</tr>
</table>
<p>The advantage of my heretical approach is that I am able to cut out a great deal of manual effort. All the domain-agnostic logic is provided by components in the framework, which means that I can spend the bulk of my time in dealing with that which has the most importance for the paying customer - the domain-specific logic. Amongst the things that I do not have to send time on are:</p>
<ol>
<li>I do not have to spend time in designing classes, methods and hierarchies for the objects in the domain/business layer as each of these represents a single table in the database. After importing a table's structure into the <a href="https://www.tonymarston.net/php-mysql/data-dictionary.html">Data Dictionary</a> subsystem in my framework I can generate the class file for that table. This is initially empty as it inherits all the invariant methods from the <a href="https://www.tonymarston.net/php-mysql/infrastructure.html#generic.table.class">abstract table class</a>. The variant or "hook" methods can be added into each table class as and when required.</li>
<li>I do not have to spend time in defining the logic which handles the basic input/output operations for each table class as all this common code is inherited from a standard <a href="https://www.tonymarston.net/php-mysql/infrastructure.html#generic.table.class">abstract table class</a>. Each of the Create, Read, Update and Delete operations is implemented as a Template Method with its own series of invariant and variant methods.</li>
<li>I do not have to spend time in writing code to validate user input before its gets written to the database as that is handled by a <a href="https://www.tonymarston.net/php-mysql/infrastructure.html#validation.class">standard framework component</a> which matches each column's value with its data type which was exported from the <a href="https://www.tonymarston.net/php-mysql/data-dictionary.html">Data Dictionary</a>.</li>
<li>I do not have to spend time in designing mechanisms to handle domain-specific logic as the methods inherited from the <a href="https://www.tonymarston.net/php-mysql/infrastructure.html#generic.table.class">abstract table class</a> already implement the <a href="https://www.tonymarston.net/php-mysql/radicore-hooks.html#template.method">Template Method Pattern</a> by providing a series of invariant methods with concrete implementations interspersed with empty <a href="https://www.tonymarston.net/php-mysql/radicore-hooks.html">hook methods</a> into which the developer can insert custom logic.</li>
<li>I do not have to spend time in designing custom Controllers which call methods which are specific to individual Model classes as the framework provides a series of reusable Controllers each of which calls generic public methods which are defined in the <a href="https://www.tonymarston.net/php-mysql/infrastructure.html#generic.table.class">abstract table class</a>. There is a different Controller provided for each <a href="https://www.tonymarston.net/php-mysql/dialog-types.html">Transaction Pattern</a>. This design allows me to pair any Controller with any Model/Table class.</li>
<li>Each event/task within the application performs a series of operations on a database table, and as there are different tasks which perform identical operations on different database tables I have managed to encapsulate the common code into a series of reusable <a href="https://www.tonymarston.net/php-mysql/dialog-types.html">Transaction Patterns</a>. A <a href="https://www.tonymarston.net/php-mysql/data-dictionary.html#generate.scripts">facility</a> within my <a href="https://www.tonymarston.net/php-mysql/data-dictionary.html">Data Dictionary</a> allows tasks to be generated by linking a particular table class with a particular pattern.</li>
<li>I do not have any Model names hard-coded into any Controller as these are passed down by each task's component script using a form of <a href="https://www.tonymarston.net/php-mysql/dependency-injection-is-evil.html#inject.model.into.the.controller">Dependency Injection</a>. Because I can now use any of my 50 Controllers with any of my 400 Models this provides an enormous amount of code reuse through <a href="https://www.tonymarston.net/php-mysql/oop-for-heretics.html#polymorphism">polymorphism</a>.</li>
<li>I do not have to spend time in designing and building individual View components for each domain object (Model) as each HTML page is constructed from a standard <a href="https://www.tonymarston.net/xml-xsl/reusable-xsl.html">reusable XSL stylesheet</a>. A single <a href="https://www.tonymarston.net/php-mysql/infrastructure.html#view.object">View</a> object, which is provided by the framework, can extract the data out of any Model and insert it into the <a href="https://www.tonymarston.net/php-mysql/infrastructure.html#xml.file">XML document</a> which is then used in the <a href="https://www.tonymarston.net/php-mysql/infrastructure.html#xsl.transformation">XSL Transformation</a> process. The differences between one web page and another have been isolated in a <a href="https://www.tonymarston.net/php-mysql/dont-do-domain-driven-design.html">screen structure file</a>
</li>
<li>I do not have to spend time in writing code to read from or write to a CSV file as the framework provides a standard component which contains all the standard code.</li>
<li>I do not have to spend time in writing code which creates PDF output as the framework provides a standard component which does this for me with the aid of a simple <a href="https://www.tonymarston.net/php-mysql/output-to-pdf.html#report.structure.file">report structure file</a> which identifies which piece of data goes where.</li>
<li>I do not have to spend time in writing code to provide <a href="https://www.tonymarston.net/php-mysql/menuguide/mnu_logon.html">user authentication</a> as the framework already provides a built-in component.</li>
<li>I do not have to spend time in writing code to construct and display <a href="https://www.tonymarston.net/php-mysql/dialog-types.html#menu-bar">menu buttons</a> as this is already <a href="https://www.tonymarston.net/php-mysql/menuguide/appendixa.html">provided by the framework</a>.</li>
<li>I do not have to spend time in writing code to construct and display <a href="https://www.tonymarston.net/php-mysql/dialog-types.html#navigation-bar">navigation buttons</a> as this is already <a href="https://www.tonymarston.net/php-mysql/menuguide/mnu_nav_button(list2).html">provided by the framework</a>.</li>
<li>I do not have to spend time in writing code to control an individual user's access to specific tasks in the application as this is already <a href="https://www.tonymarston.net/php-mysql/menuguide/appendixb.html">provided by the framework</a>.</li>
<li>I do not have to spend time in writing code to pass context between one task and another as this is already <a href="https://www.tonymarston.net/php-mysql/menuguide/appendixi.html">provided by the framework</a>.</li>
</ol>
<p>I have been building enterprise applications across numerous domains for several decades, and I have built a development framework for these applications in three different programming languages. This framework makes me very productive by providing huge volumes of domain-agnostic logic as standard from which I can then create user transactions into which I can, using my implementation of the Template Method Pattern, add domain-specific logic. I could never be as productive using an arse-backwards technique such as Domain Driven Design, so forgive me if I consign it to the toilet bowl.</p>
<p>Here endeth the lesson. Don't applaud, just throw money.</p>
<p>Here are some other heretical articles I have written on the topic of OOP:</p>
<p>Here are some articles on my framework:</p>
<hr>
<hr>
<a name="comments" id="comments"></a>
<div class="feedlyNoScript">Please enable JavaScript to view the <a href="https://disqus.com/?ref_noscript">comments powered by Disqus.</a>
</div>
<p><img border="0" alt="counter" src="http://www.tonymarston.net/counter.php?noshow&amp;page=/php-mysql/dont-do-domain-driven-design.html"></p>
</div>
</div>
</div>
<br>
<br>
<br>
<br>
<br>
<br>
via Tony Marston's Home Page<br>
<br>
December 8, 2020 at 05:54PM
